### PR TITLE
SDK's homegrown retry mechanism is inflexible & SDK isn't using request sessions

### PR DESCRIPTION
**Pull request by SteveMcGrath**
This commit should bring the SDK in-line to using more standard mechanisms to support 429s.  Further the SDK should be taking advantage of request sessions instead of using the global request mechanisms.

### CODE REVIEW

File: tenable_io/client.py [hunk 1]
```diff
@@ -46,20 +43,35 @@ def __init__(
...
         self._ini_api()
         self._init_helpers()
 
+    def _init_session(self):
+        """
+        Initializes the requests session
+        """
+        retries = Retry(
+                    total=3,
+                    status_forcelist={429},
```
Comment (alanfung) [inline tenable_io/client.py]: Currently we also retry on `501, 502, 503, 504` as well, can we add them here?

File: tenable_io/client.py [hunk 2]
```diff
@@ -46,20 +43,35 @@ def __init__(
...
+        """
+        retries = Retry(
+                    total=3,
+                    status_forcelist={429}, 
+                    backoff_factor=0.1, 
+                    respect_retry_after_header=True
+        )
+        adapter = requests.adapters.HTTPAdapter(retries)
+        self._session = requests.Session()
+        self._session.mount('https://', adapter)
```
Comment (alanfung) [inline tenable_io/client.py]: We should mount for `http://` as well to not make any assumption on how the client is configured.

File: tenable_io/client.py [hunk 3]
```diff
@@ -80,42 +92,12 @@ def _ini_api(self):
...
-                        logging.warn(u'RETRY(%d/%d)AFTER(%dms):%s' %
-                                     (count, TenableIOClient.MAX_RETRIES, sleep_ms, format_request(exception.response)))
-                        sleep(sleep_ms / 1000.0)
-                    else:
-                        raise TenableIOApiException(exception.response)
-
-        return wrapper
-
     def _error_handler(f):
         """
```
Comment (alanfung) [inline tenable_io/client.py]: Github is not allowing me to comment on the specific line, but `_error_handler` can now be simplified as there is no more need to classify error response into `TenableIORetryableApiException` (this assumes we can handle `501, 502, 503, 504` in the same manner as `429`. Moreover, `TenableIORetryableApiException` would also be obsolete.

File: tenable_io/client.py (hunk 2 above)
Comment (SteveMcGrath) [inline tenable_io/client.py]: Tenable.io's API is only interfaced over HTTPS.  we could mount this over http as well, however not sure of the value.

File: tenable_io/client.py (hunk 3 above)
Comment (SteveMcGrath) [inline tenable_io/client.py]: Yeah it simplifies a lot of things :D

File: tenable_io/client.py (hunk 1 above)
Comment (SteveMcGrath) [inline tenable_io/client.py]: I don't see any reason why not.  I was keeping it narrowly focused on 429s, but no reason not to expand it.

Comment (SteveMcGrath) [inline tenable_io/client.py]: For reference, here is the docs for Retry

http://urllib3.readthedocs.io/en/latest/reference/urllib3.util.html#module-urllib3.util.retry

File: tenable_io/client.py (hunk 2 above)
Comment (alanfung) [inline tenable_io/client.py]: If I recall correctly someone internal did in fact utilize the configurable service endpoint feature (`endpoint` in `tenable_io.ini`), however, I do not know if it was only a different hostname or different protocol all together. To answer your question, the value lies on that the SDK does not make assumption whether the API is available on HTTP or HTTPS. Perhaps it's an internal team testing against a test API instance, testing against a mock service, an outbound proxy for who-knows-what reasons... etc. I don't know and therefore I am not making an assumption. What protocol the service is available on is IMHO not an SDK-concern.

I however do not know the Tenable.io product as well as you do, so if you know for a fact that API will never be available on anything other than HTTPS. I suggest we make it impossible to configure an endpoint with HTTP to eliminate any chance an user doing so and be stuck on why the SDK not functioning correctly.

**Comment by alanfung** [comment]
I am going to port this patch and make some of the changes we've talked about. In the mean time, some of the unit tests are failing in python3, I am going to look into it. Thought I would also share it here. This is the error I am getting:

<pre>
  self = <requests.packages.urllib3._collections.RecentlyUsedContainer object at 0x1088f1048>
  key = HTTPSPoolKey(scheme='https', host='cloud.tenable.com', port=443, timeout=None, retries=None, strict=True, block=False,...ne, key_file=None, cert_file=None, cert_reqs=None, ca_certs=None, ssl_version=None, ca_cert_dir=None, ssl_context=None)
  value = <requests.packages.urllib3.connectionpool.HTTPSConnectionPool object at 0x1088ebda0>

      def __setitem__(self, key, value):
          evicted_value = _Null
          with self.lock:
              # Possibly evict the existing value of 'key'
              evicted_value = self._container.get(key, _Null)
              self._container[key] = value
      
              # If we didn't evict an existing value, we might have to evict the
              # least recently used item from the beginning of the container.
  >           if len(self._container) > self._maxsize:
  E           TypeError: unorderable types: int() > Retry()

  .venv3/lib/python3.5/site-packages/requests/packages/urllib3/_collections.py:66: TypeError
</pre>

File: tenable_io/client.py [hunk 4]
```diff
@@ -46,20 +43,35 @@ def __init__(
...
+        """
+        Initializes the requests session
+        """
+        retries = Retry(
+                    total=3,
+                    status_forcelist={429}, 
+                    backoff_factor=0.1, 
+                    respect_retry_after_header=True
+        )
+        adapter = requests.adapters.HTTPAdapter(retries)
```
Comment (alanfung) [inline tenable_io/client.py]: Leaving a note that this should have been `adapter = requests.adapters.HTTPAdapter(max_retries=retries)`

Will be porting all these changes into a new PR.

**Comment by alanfung** [comment]
Thanks for the suggestion, this is ported into PR #36.

**Pull request closed without merging**